starknet_devnet/transaction_wrapper.py

```diff
@@ -2,11 +2,102 @@
 Contains code for wrapping transactions.
 """
 
-from dataclasses import dataclass
-
-@dataclass
-class TransactionWrapper:
-    """Wraps a transaction and its receipt."""
-    def __init__(self, transaction, receipt):
-        self.transaction = transaction
-        self.receipt = receipt
+from abc import ABC, abstractmethod
+
+from starkware.starknet.definitions.error_codes import StarknetErrorCode
+from starkware.starknet.definitions.transaction_type import TransactionType
+from starkware.starknet.public.abi import get_selector_from_name
+from starkware.starknet.services.api.gateway.transaction_hash import calculate_transaction_hash
+
+from .util import fixed_length_hex
+
+
+class TransactionWrapper(ABC):
+    """Transaction Wrapper base class."""
+
+    @abstractmethod
+    def __init__(self):
+        self.transaction = {}
+        self.receipt = {}
+        self.transaction_hash = None
+
+    def generate_transaction(self, internal_transaction, status, transaction_type, **transaction_details):
+        """Creates the transaction object"""
+        self.transaction = {
+            "status": status.name,
+                "transaction": {
+                    "contract_address": fixed_length_hex(internal_transaction.contract_address),
+                    "transaction_hash": self.transaction_hash,
+                    "type": transaction_type.name,
+                    **transaction_details
+                },
+                "transaction_index": 0 # always the first (and only) tx in the block
+        }
+
+    def generate_receipt(self, execution_info):
+        """Creates the receipt for the transaction"""
+
+        self.receipt = {
+            "execution_resources": execution_info.call_info.cairo_usage,
+            "l2_to_l1_messages": execution_info.l2_to_l1_messages,
+            "status": self.transaction["status"],
+            "transaction_hash": self.transaction_hash,
+            "transaction_index": 0 # always the first (and only) tx in the block
+        }
+
+    def set_transaction_failure(self, failure_key: str, error_message: str):
```

Comment:
Don't both classes use the same failure_key, can't it be a local variable of the method instead of a function argument?